### ISSUE TITLE
Max possible fps 

### ISSUE THREAD

**Issue by k-nayak** [MRE-Lab-UMD/abd-skin-segmentation#9]
Hello @lydiazoghbi and @anirudhtopiwala  

I am using the unet_live.py code for streaming and detection in my code and model, It seems to have a lower fps and slower video when i try to stream from video file or a camera. Whereas, in the normal opencv it is very smooth. Is it something to do with the model's processing time for segmentation? What's the max achievable fps and speed of real-time stream using U-net? 
Is there any variation of Unet that is fast in realtime segmentation or is there something that i need to change in order to achieve it. I have attached my edited Unet_live code here. Any input will be appreciated. :) 

`# Importing bunch of libraries
import os
os.environ["KMP_DUPLICATE_LIB_OK"]="TRUE"
import random
import time
import datetime
import csv
import cv2
import matplotlib.pyplot as plt
import numpy as np
import torch
from skimage.io import imread, imshow
from skimage.transform import resize
from unet.unet_model import UNet
from matplotlib.animation import FuncAnimation

# Specify Image Dimensions
IMG_WIDTH = 400
IMG_HEIGHT = 400
IMG_CHANNELS = 3

seed = 42
random.seed = seed
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print("Using device: ", device)

# Load the trained model, you can use yours or the model we provide in our code
# Make sure to set up the path correctly
PATH = r"C:\Users\NAK4SI\Documents\Python_codes\Unet\Pytorch-UNet-master\checkpoints\gaus\CP_epoch50.pth"
model = UNet(n_channels=3, n_classes=1)
model.load_state_dict(torch.load(PATH, map_location=device))
model.eval()

# Open window to visualize the segmentation
cv2.namedWindow("preview")
cv2.namedWindow("Original")
#vc = cv2.VideoCapture(0, cv2.CAP_DSHOW)
vc = cv2.VideoCapture("clear.mp4")

# Try to get first frame
if vc.isOpened():
    rval, frame = vc.read()
else:
    rval = False

fps = vc.get(cv2.CAP_PROP_FPS)
start_time = time.time()
frame_count = 1
print (fps)
plt.rcParams.update({'font.size': 22})
#fourcc = cv2.VideoWriter_fourcc(*'XVID')
#out = cv2.VideoWriter('your_video.avi', fourcc, 20.0, size)
filename = r'C:\Users\NAK4SI\Documents\Python_codes\Unet\Pytorch-UNet-master\output_vid\ '+str(time.strftime("%Y-%b-%d at %H.%M.%S %p"))+'.avi'
fourcc = cv2.VideoWriter_fourcc(*'XVID')
out = cv2.VideoWriter(filename, fourcc, 30, (IMG_WIDTH, IMG_HEIGHT),0)

filename1 = r'C:\Users\NAK4SI\Documents\Python_codes\Unet\Pytorch-UNet-master\csv\ '+str(time.strftime("%Y-%b-%d at %H.%M %p"))+'.csv'  
header = ['Area %']
with open(filename1, 'a') as f:
    writer = csv.writer(f)
    writer.writerow(header)
# Loop until user presses ESC key
while rval:
    ret, frame = vc.read()
    rgb_image = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
    resized_img = np.asarray(resize(rgb_image, (IMG_HEIGHT, IMG_WIDTH), mode='constant'))
    channel_first_img = np.transpose(resized_img, (2, 0, 1))
    img_added_axis = np.expand_dims(channel_first_img, axis=0)
    input_tensor = torch.from_numpy(img_added_axis).float()
    input_tensor.to(device=device)
    preds = model(input_tensor)
    prediction = preds[0].cpu().detach().numpy()
    prediction = np.transpose(prediction, (1, 2, 0))
    # To see a binary mask, uncomment the below line.
    prediction = np.uint8((prediction > 0.7) * 255)
    #out.write(frame)
    
    img_size = np.sum(prediction>=0)#r.size
    
    white_pixels_count = np.sum(prediction==255)#np.sum(white_pixels_if_true)
    
    white_area_ratio = (white_pixels_count / img_size)*100
    frame= cv2.resize(frame,(400,400))
    cv2.imshow('Original', frame)
    area= white_area_ratio
    position = (10,50)
    cv2.putText(prediction, #numpy array on which text is written
     str(area),#text
     position, #position at which writing has to start
     cv2.FONT_HERSHEY_SIMPLEX, #font family
     1, #font size
     (240, 52, 52, 1), #font color
     3)

    cv2.imshow("preview", prediction)
    out.write(prediction)
    #out.write(frame)
    print("White pixel count : ", white_pixels_count)
    #print("size of pic : ", size)
    print("rate of fullness : % ", white_area_ratio)
    area=white_area_ratio.flatten()
    with open(filename1, 'a') as f:
        writer = csv.writer(f)
        #writer.writerow(header)
        writer.writerow(area)
    x= ((time.time()-start_time)/60)
    y=area   
    plt.scatter(x, y)
    plt.title("Real Time plot")
    plt.xlabel("Time (minutes)")
    plt.ylabel("Area covered (%)")
    plt.pause(1)

    key = cv2.waitKey(20)
    if key == 27:  # exit on ESC
        break
    '''if frame_count % 30 == 0:
        print("Frame Per second: {} fps.".format(
            (time.time() - start_time) / frame_count))
    frame_count = frame_count + 1'''
#plt.show()
vc.release()
out.release()
cv2.destroyWindow("preview")
cv2.destroyWindow("Original")
cv2.destroyAllWindows()     

# Close window
`

**Comment by lydiazoghbi** [comment]
This is likely a very late response but I'll put it in here since it may help others. Just quickly glancing at what you pasted, your image size is over 3 times larger than what we have trained our network on. I am going to assume that you have trained your own network because ours will not return anything reasonable with a 400 x 400 image input, in which case I still cannot reasonably comment on the lags without knowing your network architecture. Streaming images without any post-processing operations is one thing, and running a network on a real-time video stream is a totally different thing. It can also depend on the machine you're using. In other words, there's a lot of variables here, and you'd have to narrow down your question and provide additional information to pinpoint the source of the lag you're experiencing. Last we tried in 2019, we got an fps of 37.
